### ISSUE TITLE
Decoding event data broken if data length is not a multiple of 32

### ISSUE THREAD

**Issue by Jannis** [paritytech/ethabi#139]
## Problem description

Livepeer's `JobsManager` contract emits `NewJob(address,uint256,string,string,uint256,uint256)` events. Most (if not all) if these events have a data length that is not a multiple of 32. Here are two examples:

* https://etherscan.io/tx/0x712ed897c42e20de0f369eaaea32acc85fa9dd5f88e93170f64e00c9351cbc4a#eventlog
* https://etherscan.io/tx/0xc8d552532338e9ee470c1912d42c2fda0b41bdc1d8aebd5831ff43abcf6a24fd#eventlog

Trying to decode the event params with e.g.
```rust
let encoded = hex!(
    "
     0000000000000000000000000000000000000000000000000000000000000000
     00000000000000000000000000000000000000000000000000000000000000a0
     0000000000000000000000000000000000000000000000000000000000000152
     0000000000000000000000000000000000000000000000000000000000000001
     000000000000000000000000000000000000000000000000000000000054840d
     0000000000000000000000000000000000000000000000000000000000000092
     3132323033393637623533326130633134633938306235616566666231373034
     3862646661656632633239336139353039663038656233633662306635663866
     3039343265376239636337366361353163636132366365353436393230343438
     6533303866646136383730623565326165313261323430396439343264653432
     3831313350373230703330667073313678390000000000000000000000000000
     0000000000000000000000000000000000103933633731376537633061363531
     3761
"
);

let decoded = decode(
    &[
        ParamType::Uint(256),
        ParamType::String,
        ParamType::String,
        ParamType::Uint(256),
        ParamType::Uint(256),
    ],
    &encoded,
).unwrap()
```
currently fails because `util::slice_data` expects the length of the encoded data to be  multiple of 32.

## Web3.js behavior

Web3.js parses the same data fine, as the following log from the Node.js shell demonstrates:
```js
> const Web3 = require('web3');
> const web3 = new Web3();
> web3.eth.abi.decodeParameters(['uint256', 'string', 'string', 'uint256', 'uint256'], '0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000001520000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000054840d00000000000000000000000000000000000000000000000000000000000000923132323033393637623533326130633134633938306235616566666231373034386264666165663263323933613935303966303865623363366230663566386630393432653762396363373663613531636361323663653534363932303434386533303866646136383730623565326165313261323430396439343264653432383131335037323070333066707331367839000000000000000000000000000000000000000000000000000000000000001039336337313765376330613635313761')
Result {
  '0': '0',
  '1':
   '12203967b532a0c14c980b5aeffb17048bdfaef2c293a9509f08eb3c6b0f5f8f0942e7b9cc76ca51cca26ce546920448e308fda6870b5e2ae12a2409d942de428113P720p30fps16x9',
  '2': '93c717e7c0a6517a',
  '3': '1',
  '4': '5538829',
  __length__: 5 }
```

## Desired behavior

The desired behavior would be for `ethabi` to be able to decide the event data in the same way `web3.js` does it.

**Comment by debris** [comment]
```
     0000000000000000000000000000000000000000000000000000000000000000
     00000000000000000000000000000000000000000000000000000000000000a0
     0000000000000000000000000000000000000000000000000000000000000152
     0000000000000000000000000000000000000000000000000000000000000001
     000000000000000000000000000000000000000000000000000000000054840d
     0000000000000000000000000000000000000000000000000000000000000092
     3132323033393637623533326130633134633938306235616566666231373034
     3862646661656632633239336139353039663038656233633662306635663866
     3039343265376239636337366361353163636132366365353436393230343438
     6533303866646136383730623565326165313261323430396439343264653432
     3831313350373230703330667073313678390000000000000000000000000000
     0000000000000000000000000000000000103933633731376537633061363531
     3761
```

### this is an invalid encoding

----

event `NewJob`

```
NewJob(address indexed broadcaster, uint256 jobId, string streamId, string transcodingOptions, uint256 maxPricePerSegment, uint256 creationBlock);
```

has 2 variadic size arguments - strings. According to the solidity abi specification they should be aligned to the multiple of 32 bytes.

https://solidity.readthedocs.io/en/develop/abi-spec.html#use-of-dynamic-types

TL;DR

The problem is not in `ethabi`, but in solidity compiler. Please report it upstream

**Comment by Jannis** [comment]
I can, but contracts with this—presumably broken—encoding exist on mainnet, and I'd argue that they need to be supported.

**Comment by Jannis** [comment]
Another thought here: Which component decides how events are encoded and written to the blockchain? Is this really a bug in solidity or could it come from e.g. parity / go-ethereum (depending on what the node was running that executed the transaction and put this data into the block)?

**Comment by dvdplm** [comment]
Closing this issue as, while the concern is certainly valid, it is out of scope for `ethabi` to handle invalid encodings.
